lib/datadog/grok/src/field_traversal.rs

```diff
@@ -0,0 +1,120 @@
+use lookup::{Field, FieldBuf, Lookup, LookupBuf, Segment, SegmentBuf};
+use parsing::value::Value;
+use snafu::Snafu;
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+
+pub fn get_field<'a>(
+    source: &Value,
+    lookup: impl Into<Lookup<'a>> + Debug,
+) -> std::result::Result<Option<&Value>, Error> {
+    let mut working_lookup = lookup.into();
+
+    let this_segment = working_lookup.pop_front();
+    match (this_segment, source) {
+        // We've met an end and found our value.
+        (None, item) => Ok(Some(item)),
+        // Descend into a map
+        (Some(Segment::Field(Field { name, .. })), Value::Map(map)) => match map.get(name) {
+            Some(inner) => get_field(inner, working_lookup.clone()),
+            None => Ok(None),
+        },
+        // Descend into an array
+        (Some(Segment::Index(i)), Value::Array(array)) => match array.get(i as usize) {
+            Some(inner) => get_field(inner, working_lookup.clone()),
+            None => Ok(None),
+        },
+        // anything else is not allowed
+        _ => Err(Error::LookupFailed {
+            lookup: working_lookup.into_buf(),
+        }),
+    }
+}
+
+pub fn insert_field(
+    source: &mut Value,
+    lookup: impl Into<LookupBuf> + Debug,
+    value: Value,
+) -> std::result::Result<Option<Value>, Error> {
```

Comment:
Does this need to be a `std::result::Result<Option<Value>, Error>`?

It seems like each branch swaps the `source` + `value` in place, and in each case where `insert_field` is called, the return value is ignored.

If this is being used for the Result ergonomics, might it be better represented as a `std::result::Result<(), Error>`?